File: ConversationIDE/src/VanEngine.Core/VAN/metrics.py
from __future__ import annotations

import threading
import time
from typing import Any


class Metrics:
    def __init__(self):
        self._envelopes_processed = 0
        self._errors_encountered = 0
        self._total_processing_ticks = 0
        self._active_processors = 0
        self._lock = threading.Lock()

    @property
    def envelopes_processed(self) -> int:
        return self._envelopes_processed

    @property
    def errors_encountered(self) -> int:
        return self._errors_encountered

    @property
    def total_processing_ticks(self) -> int:
        return self._total_processing_ticks

    @property
    def active_processors(self) -> int:
        return self._active_processors

    @property
    def average_processing_ms(self) -> float:
        count = self.envelopes_processed
        return (self.total_processing_ticks / count / 10_000) if count > 0 else 0.0

    def begin_processing(self) -> _MetricsScope:
        with self._lock:
            self._active_processors += 1
        return _MetricsScope(self)

    def record_envelope(self) -> None:
        with self._lock:
            self._envelopes_processed += 1

    def record_error(self) -> None:
        with self._lock:
            self._errors_encountered += 1

    def record_ticks(self, ticks: int) -> None:
        with self._lock:
            self._total_processing_ticks += ticks

    def snapshot(self) -> dict[str, Any]:
        return {
            "envelopes_processed": self.envelopes_processed,
            "errors_encountered": self.errors_encountered,
            "average_processing_ms": self.average_processing_ms,
            "active_processors": self.active_processors,
            "total_processing_ticks": self.total_processing_ticks,
        }

    def reset(self) -> None:
        with self._lock:
            self._envelopes_processed = 0
            self._errors_encountered = 0
            self._total_processing_ticks = 0
            self._active_processors = 0


class _MetricsScope:
    def __init__(self, metrics: Metrics):
        self._metrics = metrics
        self._start = time.perf_counter_ns()

    def __enter__(self) -> _MetricsScope:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        elapsed_ns = time.perf_counter_ns() - self._start
        elapsed_ticks = elapsed_ns  # 1 tick = 100ns in .NET; use ns directly
        self._metrics.record_ticks(elapsed_ticks)
        with self._metrics._lock:
            self._metrics._active_processors -= 1

File: src/VanEngine.Core/VAN/Metrics.cs
using System.Diagnostics;
using System.Threading;

namespace VanEngine.Core.VAN;

public sealed class Metrics
{
    private long _envelopesProcessed;
    private long _errorsEncountered;
    private long _totalProcessingTicks;
    private int _activeProcessors;

    public long EnvelopesProcessed => Interlocked.Read(ref _envelopesProcessed);
    public long ErrorsEncountered => Interlocked.Read(ref _errorsEncountered);
    public long TotalProcessingTicks => Interlocked.Read(ref _totalProcessingTicks);
    public int ActiveProcessors => Thread.VolatileRead(ref _activeProcessors);

    public double AverageProcessingMs
    {
        get
        {
            long count = EnvelopesProcessed;
            return count > 0 ? (double)TotalProcessingTicks / count / TimeSpan.TicksPerMillisecond : 0;
        }
    }

    public IDisposable BeginProcessing()
    {
        Interlocked.Increment(ref _activeProcessors);
        return new MetricsScope(this, Stopwatch.StartNew());
    }

    public void RecordEnvelope()
    {
        Interlocked.Increment(ref _envelopesProcessed);
    }

    public void RecordError()
    {
        Interlocked.Increment(ref _errorsEncountered);
    }

    public void RecordTicks(long ticks)
    {
        Interlocked.Add(ref _totalProcessingTicks, ticks);
    }

    public Dictionary<string, object> Snapshot()
    {
        return new()
        {
            ["envelopes_processed"] = EnvelopesProcessed,
            ["errors_encountered"] = ErrorsEncountered,
            ["average_processing_ms"] = AverageProcessingMs,
            ["active_processors"] = ActiveProcessors,
            ["total_processing_ticks"] = TotalProcessingTicks
        };
    }

    public void Reset()
    {
        Interlocked.Exchange(ref _envelopesProcessed, 0);
        Interlocked.Exchange(ref _errorsEncountered, 0);
        Interlocked.Exchange(ref _totalProcessingTicks, 0);
        Interlocked.Exchange(ref _activeProcessors, 0);
    }

    private sealed class MetricsScope : IDisposable
    {
        private readonly Metrics _metrics;
        private readonly Stopwatch _sw;

        public MetricsScope(Metrics metrics, Stopwatch sw)
        {
            _metrics = metrics;
            _sw = sw;
        }

        public void Dispose()
        {
            _sw.Stop();
            _metrics.RecordTicks(_sw.ElapsedTicks);
            Interlocked.Decrement(ref _metrics._activeProcessors);
        }
    }
}
